data_structures/src/fee.rs

```diff
@@ -0,0 +1,205 @@
+}
+
+impl num_traits::Zero for RelativeFee {
+    #[inline]
+    fn zero() -> Self {
+        Self(Priority::zero())
+    }
+
+    #[inline]
+    fn is_zero(&self) -> bool {
+        self.0.is_zero()
+    }
+}
+
+#[derive(Copy, Clone, Debug, Deserialize, Hash, PartialEq, Eq, PartialOrd, Serialize)]
+/// Type for representing a fee value that can be absolute (nanoWits) or relative (priority).
+pub enum Fee {
+    /// An absolute fee, as expressed in nanoWits.
+    Absolute(AbsoluteFee),
+    /// A relative fee, aka "priority", as expressed as nanoWits (or fractional amounts) per weightunit.
+    Relative(RelativeFee),
+}
+
+impl Fee {
+    #[inline]
+    pub fn absolute_from_wit(wit: Wit) -> Self {
+        Self::Absolute(AbsoluteFee(wit))
+    }
+
+    #[inline]
+    pub fn absolute_from_nanowits(nanowits: u64) -> Self {
+        Self::absolute_from_wit(Wit::from_nanowits(nanowits))
+    }
+
+    pub fn relative_from_float<T>(float: T) -> Self
+    where
+        f64: From<T>,
+    {
+        Self::Relative(RelativeFee(Priority::from(f64::from(float))))
+    }
+}
+
+impl Default for Fee {
+    fn default() -> Self {
+        <Self as Zero>::zero()
+    }
+}
+
+impl fmt::Display for Fee {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Fee::Absolute(absolute) => absolute.fmt(f),
+            Fee::Relative(relative) => relative.fmt(f),
+        }
+    }
+}
+
+impl ops::Add for Fee {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        use crate::fee::Fee::*;
+        match (self, rhs) {
+            (Absolute(lhs), Absolute(rhs)) => Fee::Absolute(lhs + rhs),
+            (Relative(lhs), Relative(rhs)) => Fee::Relative(lhs + rhs),
+            _ => {
+                unimplemented!()
+            }
+        }
+    }
+}
+
+impl num_traits::Zero for Fee {
+    #[inline]
+    fn zero() -> Self {
+        Self::Absolute(AbsoluteFee::zero())
+    }
+
+    fn is_zero(&self) -> bool {
+        match self {
+            Fee::Absolute(absolute) => absolute.is_zero(),
+            Fee::Relative(relative) => relative.is_zero(),
+        }
+    }
+}
+
+/// Allow backwards compatibility with old Wallet API clients that may provide fee values without
+/// tagging whether they are absolute or relative.
+///
+/// This implicitly treats integers as absolute fees and floats as relative fees. Strings encoding
+/// numbers are also parsed in the same way.
+pub fn deserialize_fee_backwards_compatible<'de, D>(deserializer: D) -> Result<Fee, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    #[derive(Deserialize)]
+    #[serde(untagged)]
+    enum Untagged {
+        Fee(Fee),
+        Float(f64),
+        Integer(u64),
+        String(String),
+    }
+
+    Ok(match Untagged::deserialize(deserializer)? {
+        Untagged::Fee(fee) => fee,
+        Untagged::Float(float) => Fee::relative_from_float(float),
+        Untagged::Integer(integer) => Fee::absolute_from_nanowits(integer),
```

Comment:
This distinction between absolute and relative doesn't look like a good idea. It's true that absolute fees must always be integers, but relative fees can be integers or floats. In fact, in the master branch they are always integers so I'm not sure if this will work as expected.

Reply:
Any suggestions on how to overcome that limitation?

Reply:
Since this is only used in the wallet, we can break backwards compatibility and make sure the next version of sheikah uses this version of the wallet. Also we would need to make sure that the current version of sheikah does not update the wallet, as that would break it.

Reply:
> In fact, in the master branch they are always integers so I'm not sure if this will work as expected.

Actually, in the master branch they are always integers encoded as strings, so only the string branch will be entered by old clients.

Reply:
> Since this is only used in the wallet, we can break backwards compatibility and make sure the next version of sheikah uses this version of the wallet. Also we would need to make sure that the current version of sheikah does not update the wallet, as that would break it.

Actually we cannot because exchanges use the wallet too, and afaik they use absolute fees in integer numbers, hence this way of approaching backwards compatibility :crying_cat_face: